runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java

```diff
@@ -361,7 +362,18 @@ public void setup() {
     public void processElement(final ProcessContext c) {
       final InputT element = c.element().getKey();
       invoker.invokeSplitRestriction(
-          element, c.element().getValue(), part -> c.output(KV.of(element, part)));
+          element, c.element().getValue(), new OutputReceiver<RestrictionT>() {
+            @Override
+            public void output(RestrictionT part) {
+              c.output(KV.of(element, part));
+            }
+
+            @Override
+            public void outputWithTimestamp(RestrictionT part, Instant timestamp) {
+              c.outputWithTimestamp(KV.of(element, part), timestamp);
```

Comment:
<!--thread_id:cc_183501107_t; commit:e7163d592d424a480ae9071c770022129d401905; resolved:0-->
<!--section:context-quote-->
> **jkff** wrote:
> I think this one should be an UnsupportedOperationException - it probably doesn't make sense to have different splits of a restriction have different timestamps.

<!--section:body-->
Can you explain? This is the output element, right? is there a reason not to allow outputting with a timestamp?

Reply:
This is not an output element, but an output restriction - we're splitting a restriction into several sub-restrictions adding up to the original. I don't think there's ever a case for splitting (E, R)@t0 into (E, R1)@t1, (E, R2)@t2 where t1 or t2 are different from t0 - the timestamp belongs to the element, not the restrictions.

Reply:
As a tie-breaker: the dynamic splitting API does not allow you to assign different timestamps to the restrictions, and neither should the static one.